Add support for ARM+JTAG for JLink probes

thefaxman opened the pull request:
This is the first step towards supporting ARMv7-A, which often is JTAG only.  Summary of changes:

* Add methods to allow callers to inspect what protocol a probe is currently using
* Update ARM usage of swj_sequence to select between JTAG and SWD sequences
* Update JLink probe implementation to handle both JTAG and SWD transactions
* Generalize JLink code where possible to share between SWD and JTAG
* Basic tests for JTAG operations

In addition to the tests this has been verified on real hardware with a JLink connected to a Raspberry Pi 4 and several STM32 devices to test both SWD and JTAG.

Signed-off-by: Ryan Fairfax <ryan@thefaxman.net>

noppej commented:
This is awesome. When connecting to RPi 4, did you connect via USB or was their special wiring involved?

thefaxman commented:
> This is awesome. When connecting to RPi 4, did you connect via USB or was their special wiring involved?

The JTAG pins are exposed via the 40 pin GPIO header and a config.txt setting to enable JTAG.  You have to manually hook the wires up to the probe but it's not too hard to get setup.  This is a good reference for OpenOCD which contains the wiring diagram: https://linaro.atlassian.net/wiki/spaces/TCWGPUB/pages/25296346120/Raspberry+Pi+Linux+kernel+debugging+with+OpenOCD

Note that with the Raspberry Pi this is enough to get the DAP / AP operations over JTAG working and the ROM table is properly read, but CPU halt / step / inspect / etc. isn't quite there yet because we don't have Cortex-A support.  That will be next on my list once this goes in.

thefaxman commented:
One specific thing I'm looking for feedback on - I added set_ir_len() to the probe trait.  This is to work around hard coded assumptions of irlen currently in drivers - ADIv5 is always 4 and the RISC-V support assumes 5.  You can see with the old JLink code it assumed a length of 5 because it only supported RISC-V operations over JTAG.

This is a stop gap as IR Lengths are really per TAP, not per probe.  However, most probes don't expose TAPs and there's no external to the probe API surface.  In practice, most ARM devices only have one TAP.

I added the operation to let the RISC-V / ARM code set the expected IR length, but I'm a little uneasy about it as I'm not sure it's the right shape long term for things like multi-tap support.

If folks have opinions on how to better model this I'd love to hear them.

Tiwalun commented on probe-rs/src/probe.rs:
```diff
@@ -802,6 +805,9 @@ pub trait JTAGAccess: DebugProbe {
     /// Return the currently configured idle cycles.
     fn get_idle_cycles(&self) -> u8;
 
+    // Set the IR register length
```
```suggestion
    /// Set the IR register length
```

Tiwalun commented on probe-rs/src/probe/jlink/swd.rs:
```diff
@@ -122,20 +123,217 @@ impl SwdStatistics {
+    let received = parse_jtag_response(&result);
+
+    if transfer.is_abort() {
+        // No resposnes returned from this
```
```suggestion
        // No responses returned from this
```

Tiwalun commented on probe-rs/src/probe/jlink/swd.rs:
```diff
@@ -122,20 +123,217 @@ impl SwdStatistics {
+
+    // This is a bit confusing, but a read from any port is still
+    // a JTAG write as we have to transmit the address
+    let result = probe.write_register(address, &data[..], 35)?;
```
What does the 35 mean here?

Tiwalun commented on probe-rs/src/probe/jlink/swd.rs:
```diff
@@ -122,20 +123,217 @@ impl SwdStatistics {
+            TransferStatus::Failed(DapError::WaitResponse)
+        }
+        JTAG_STATUS_OK => {
+            // OK/FAULT
```
Is there a separate status for FAULT?

thefaxman replied:
No - and this is a bit confusing.  The value in the ARM JTAG documentation is described as OK/FAULT.

There's not a separate fault bit as unlike SWD there are no protocol level failures.  A 0/1 is always valid for every bit.  If you read an invalid register, such as one that is SWD only, in JTAG it will return OK/FAULT and an undefined value.  You're basically supposed to ensure your JTAG code doesn't make bad requests, which makes debugging a bit hard but there's not much more we can do here.

The only fault that can actually happen assuming you're reading valid registers is an overrun, which returns WAIT.

Tiwalun reviewed (commented):
Thanks for working on this!

Some small nits, but looks good otherwise. 

Maybe we should change the name of the `jlink/swd.rs` file, since it now does not only SWD, but JTAG as well. I don't think there is a proper name, but maybe call it something like `arm_debug.rs`?

I think the `set_ir_len`is good enough for now, in the long term we should probably have a proper interface to specify TAPs, and then we can use that for ARM and RISCV.

thefaxman commented:
> Maybe we should change the name of the `jlink/swd.rs` file, since it now does not only SWD, but JTAG as well. I don't think there is a proper name, but maybe call it something like `arm_debug.rs`?

Great suggestion, updated in the latest commits.

Yatekii commented on probe-rs/src/architecture/arm/sequences/mod.rs:
```diff
@@ -91,9 +91,22 @@ pub trait ArmDebugSequence: Send + Sync {
+                interface.swj_sequence(16, 0xE79E)?;
+            }
+            _ => {
+                return Err(crate::Error::Probe(DebugProbeError::NotImplemented(
```
I am not sure if it's correct to fail here. Can't the `SWJ-DP Switch Sequence` always be executed for example, I thought so. If the probe would be stuck in unknown state for whatever reason and we cannot read the current protocol, it would be stuck there forever.

thefaxman replied:
The idea is that if we don't have a protocol selected there is no way to generate a correct SWJ sequence as we don't know what we're switching to.  The only alternative I could think to do is to default to either SWD or JTAG if not specified.

We could panic as getting here without a protocol error selected will never work / recover and is a usage error.

noppej replied:
> We could panic as getting here without a protocol error selected will never work / recover and is a usage error.

@thefaxman I would prefer it if we never panic. VSCode users, via the probe-rs-debugger, won't see panics. I would prefer if we can propogate a useful error that the UX can catch and help the user be more successful next time :)

thefaxman replied:
> I would prefer it if we never panic.

I think that makes sense, the question is then should the current logic that returns an error remain or not.  The only other option I could see to do is to default to a protocol, but I think that's the choice of the probe and not of this layer as if they disagree this will just break.

Yatekii replied:
Hmm that's fair. The name is missleading then I think. `active_protocol` - to me - suggests that it's the currently active protocol, you are switching **from** not **to**.
So the code makes sense to me, maybe the name not so much in this context. In general it makes sense tho.
So let's leave as is.

thefaxman replied:
I can add a comment here indicating that the goal is to ensure the DAP is speaking the protocol selected via active_protocol, regardless of any current DAP state.  Would that help clarify the goal?

Yatekii replied:
Yeah that would have been good, but as always I was a bit trigger happy. Can do it retroactively if you feel like :)

thefaxman replied:
No problem - I'll put the comment in my next batch of changes.  I'd rather commit early and often.

Yatekii reviewed (approved):
Thanks a lot for this :)

This was really quick!

bors r+
